### PR TITLE
Add EventListener Selector For TriggerCRD

### PR DESCRIPTION
**Pull request by khrm**
# Changes
EventListener have NamespaceSelector field which gives us namespaces from where EventListener fetches Trigger object to process events.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes
```release-note

Added Namespace Selector field for EventListener which enables EventListener to serve across the namespace. `namespaceSelector` field with `matchNames` need to be provided to enable selector.
`
  namespaceSelector:
    matchNames:
    - nsName1
    - nsName2
`
```

### CODE REVIEW

**Comment by dibyom** [comment]
@khrm was there additional work you were planning on doing before unmarking this as WIP?

(One thing we should add are some release notes and some docs/examples).

File: pkg/apis/triggers/v1alpha1/event_listener_types.go [hunk 1]
```diff
@@ -152,6 +153,17 @@ type EventListenerConfig struct {
 	GeneratedResourceName string `json:"generatedName"`
 }
 
+// NamespaceSelector is a selector for selecting either all namespaces or a
+// list of namespaces.
+// +k8s:openapi-gen=true
+type NamespaceSelector struct {
+	// Boolean describing whether all namespaces are selected in contrast to a
```
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/event_listener_types.go:160]: AdmissionWebhooks also have NameSpaceSelector (and ObjectSelector). We could take some pointers from there: https://github.com/kubernetes/api/blob/master/admissionregistration/v1beta1/types.go#L410


(might be worth seeing if there are any reusable bits in the implementation that we could use too)

[the remark this answers is not in the document]
Comment (khrm) [reply pkg/apis/triggers/v1alpha1/event_listener_types.go:159]: Yes, `any` here could be dispense with but AdmissionWebhook implementation is for more complex scenarios. https://github.com/kubernetes/api/blob/8519c5ea46199d57724725d5b969c5e8e0533692/admissionregistration/v1beta1/types.go#L372

File: pkg/sink/initialization_test.go [hunk 2]
```diff
@@ -31,6 +33,10 @@ func Test_GetArgs(t *testing.T) {
 	if err := flag.Set(port, "port"); err != nil {
 		t.Errorf("Error setting flag port: %s", err)
 	}
+	if err := flag.Set(elType, v1alpha1.ELTypeNamespaced); err != nil {
+		t.Errorf("Error setting flag port: %s", err)
```
Comment (dibyom) [inline pkg/sink/initialization_test.go:37]:
```suggestion
		t.Errorf("Error setting flag elType: %s", err)
```

File: pkg/sink/initialization_test.go [hunk 3]
```diff
@@ -44,6 +50,9 @@ func Test_GetArgs(t *testing.T) {
 	if sinkArgs.Port != "port" {
 		t.Errorf("Error port want port, got %s", sinkArgs.Port)
 	}
+	if sinkArgs.ElType != v1alpha1.ELTypeNamespaced {
+		t.Errorf("Error EL Type want type, got %s", sinkArgs.Port)
```
Comment (dibyom) [inline pkg/sink/initialization_test.go:54]:
```suggestion
		t.Errorf("Error EL Type want type, got %s", sinkArgs.ElType)
```

File: pkg/sink/initialization.go [hunk 4]
```diff
@@ -51,6 +54,8 @@ type Args struct {
 	ElNamespace string
 	// Port is the port the Sink should listen on.
 	Port string
+	// ELType determines whether el functions as namespaced or clustered
```
Comment (dibyom) [inline pkg/sink/initialization.go:57]: minor: would be clearer to make this a boolean flag e.g. IsClusterWide or IsMultiNameSpace

File: examples/selectors/01_rbac.yaml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: foo-el-sa
+  namespace: foo
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: tekton-triggers-example-binding
```
[the remark this answers is not in the document]
Comment (khrm) [reply examples/selectors/01_rbac.yaml]: I change the name to better reflect that its for el.

File: pkg/sink/initialization.go [hunk 6]
```diff
@@ -50,6 +51,8 @@ var (
 		"The idle timeout for EventListener Server.")
 	elTimeOutHandler = flag.Int64("timeouthandler", 5,
 		"The timeout for Timeout Handler of EventListener Server.")
+	isMultiNSFlag = flag.String("is-multi-ns", "",
```
Comment (dibyom) [inline pkg/sink/initialization.go:54]:
```suggestion
	isMultiNSFlag = flag.Bool("is-multi-ns", false, 
```

**Review by dibyom — COMMENTED** [review]
One minor nit. But otherwise LGTM!

/approve

**Comment by savitaashture** [comment]
@khrm can we add e2e to use trigger with NSSelector?

File: examples/selectors/03_trigger.yaml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        value: $(tt.params.contenttype)
+      resources:
+      - name: git-source
+        resourceSpec:
+          type: git
+          params:
+          - name: revision
+            value: $(tt.params.gitrevision)
+          - name: url
+            value: $(tt.params.gitrepositoryurl)
```
Comment (savitaashture) [inline examples/selectors/03_trigger.yaml:77]: Can we add README in selector folder to setup locally like github example

File: examples/selectors/02_eventlistener-sel.yaml [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+kind: EventListener
+metadata:
+  name: listener-sel
+  namespace: foo
+spec:
+  serviceAccountName: foo-el-sa
+  namespaceSelector:
+    matchNames:
+    - foo
+    - bar
```
Comment (savitaashture) [inline examples/selectors/02_eventlistener-sel.yaml:11]: If we specify 2 ns here does it mean EL will search in both the namespace for triggers and its not mandatory to have `trigger` in both the namespace am i right (I mean `trigger` can present in any one of the ns)

What happens if no `trigger` exist in both the ns?

File: pkg/sink/sink_test.go [hunk 9]
```diff
@@ -270,6 +270,115 @@ func TestHandleEvent(t *testing.T) {
...
+			bldr.TriggerBindingParam("revision", "$(body.head_commit.id)"),
+			bldr.TriggerBindingParam("foo", "$(body.foo)"),
+			bldr.TriggerBindingParam("type", "$(header.Content-Type)"),
+		))}
+	// Add TriggerBinding to Trigger
+	triggers := []*triggersv1.Trigger{bldr.Trigger("my-trigger", triggerNS, bldr.TriggerSpec(
+		bldr.TriggerSpecBinding("my-triggerbinding", "", "my-triggerbinding", "v1alpha1"),
+		bldr.TriggerSpecTemplate("my-triggertemplate", "v1alpha1"),
+	))}
+	// Add TriggerRef to EventListener
```
Comment (savitaashture) [inline pkg/sink/sink_test.go:327]: I think we are not adding triggerRef here right :thinking:

File: examples/selectors/02_eventlistener-sel.yaml (hunk 8 above)
Comment (khrm) [reply examples/selectors/02_eventlistener-sel.yaml:11]: Yes. If no trigger is present, then nothing happens.

Comment (khrm) [reply examples/selectors/02_eventlistener-sel.yaml:11]: We wait for a user to create trigger.

File: examples/selectors/03_trigger.yaml (hunk 7 above)
Comment (khrm) [reply examples/selectors/03_trigger.yaml:77]: Done,

**Comment by khrm** [comment]
@savitaashture  I would create another pr for e2e test later today.

**Review by savitaashture — COMMENTED** [review]
/lgtm

**Pull request merged**